crates/wysiwyg/src/composer_model/format.rs

```diff
@@ -206,46 +206,59 @@ where
         range: &Range,
         format: &InlineFormatType,
     ) {
-        for location in range.locations.iter() {
-            let node = self.state.dom.lookup_node(&location.node_handle);
-            if let DomNode::Container(n) = node {
-                if let ContainerNodeKind::Formatting(f) = n.kind() {
-                    if f == format {
-                        if node.has_only_placeholder_text_child() {
-                            self.state.end = self.state.start;
-                            self.state
-                                .dom
-                                .replace(&location.node_handle, vec![]);
-                        } else {
-                            self.state.dom.remove_and_keep_children(
-                                &location.node_handle,
-                            );
-                        }
+        // Filter locations for formatting nodes.
+        let formatting_locations: Vec<&DomLocation> = range
+            .locations
+            .iter()
+            .filter(|l| {
+                let n = self.state.dom.lookup_node(&l.node_handle);
+                n.is_formatting_node_of_type(format)
+            })
+            .rev()
+            .collect();
+
+        // Find slices of text before and after the selection that will require re-format.
+        let mut reformat_to: Option<usize> = None;
+        let mut reformat_from: Option<usize> = None;
+        if let Some(location) = formatting_locations.first() {
+            // Actual last node, find text to reformat after.
+            let (after_start, after_end) = self.safe_locations_from(
+                Location::from(end),
+                Location::from(end + location.length - location.end_offset),
+            );
```

Comment:
Wonder if we really need the `safe_locations_from` here, as if the model and the find_range work properly we'll never have invalid locations (perhaps it's better to crash than having weird behaviour if it ever happens)